src/nvim/terminal.c

```diff
@@ -2042,23 +2021,35 @@ static void refresh_terminal(Terminal *term)
   }
   linenr_T ml_before = buf->b_ml.ml_line_count;
 
-  // refresh_ functions assume the terminal buffer is current
+  // refresh_ functions assume the terminal buffer is current. Don't call refresh_cursor here, as we
+  // don't want a terminal that was possibly made temporarily current influencing the cursor.
   aco_save_T aco;
   aucmd_prepbuf(&aco, buf);
   refresh_size(term, buf);
   refresh_scrollback(term, buf);
   refresh_screen(term, buf);
-  refresh_cursor(term);
   aucmd_restbuf(&aco);
 
   int ml_added = buf->b_ml.ml_line_count - ml_before;
   adjust_topline(term, buf, ml_added);
 }
 
-static void refresh_cursor(Terminal *term)
+static void refresh_cursor(Terminal *term, bool *cursor_visible)
```

Comment:
Since `refresh_cursor` is no longer called in `refresh_terminal` and is only called in `terminal_check`, would it make sense to move/combine this function into `terminal_check_cursor`?